### PR TITLE
Add welcome popup

### PR DESCRIPTION
**Pull request by dffdff2423**
## About the PR

Add a popup window that explains the basics of CD. It has an option to not show the current revision of the window again. 

## Why / Balance

Nairod asked on discord.

## Technical details

We hijack the document parser that the guidebook uses so that it is possible to embed entities in the window without modifying code. Upon clicking the "do not show again" button an integer representing the current revision of the window is written to a file inside the user data directory. We check this before opening the window on roundstart.

Please remember to update the revision number inside `Content.Client/_CD/WelcomeMessage/WelcomeWindowUIController.cs` whenever a major change to the message contents is made.

## Media

image

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: Aquif
- add: Add a welcome message so people don't need to ask how to get to the station anymore.

### CODE REVIEW

File: Resources/ServerInfo/InGameWelcome.xml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+See the gudebook for more information.
+
+<Box>
+<GuideEntityEmbed Entity="MachineCryoSleepPod" />
+</Box>
+
+Lastly, if you need something from admins [color=#FDDA0D]please don't hesitate to ahelp[/color].
+This is not limited to rules related things. The admins are happy to modify your
+character, give you an antagonist role, and do many other things. The worst they can
```
[the remark this answers is not in the document]
Comment (dffdff2423) [reply Resources/ServerInfo/InGameWelcome.xml:29]: I think the yellow stands out a bit more, but I will wait for more feedback. If the admins get too many ahelps about antags we can always change it.

**Review by Nairodian — APPROVED** [review]
Can't say much on the actual code, but aside from one or two tiny nitpicks, info's all good.

File: Content.Client/_CD/WelcomeMessage/WelcomeWindowUIController.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    // Please try to remember to increment this when the file below is changed. Sadly, there is no way
+    // to leave a comment in the scuffed XML-robust markup hybrid ss14 uses.
+    private const int TextRevision = 1;
+    private const string ContentsPath = "/ServerInfo/InGameWelcome.xml";
+
+    private const string LastSeenRevisionPath = "/cosmatic_welcome_last_seen";
+
+    private WelcomeWindow? _window;
+
```
Comment (metalgearsloth) [inline Content.Client/_CD/WelcomeMessage/WelcomeWindowUIController.cs:24]: Can't you just hash the file contents and store that in the cvar?

File: Content.Client/_CD/WelcomeMessage/WelcomeWindow.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        using var file = _resourceManager.ContentFileReadText(path);
+
+        if (!_parsingMan.TryAddMarkup(Contents, file.ReadToEnd()))
+        {
+            Contents.AddChild(new Label() { Text = "ERROR: Failed to parse document." });
+            _sawmill.Error($"Failed to parse contents of guide document {path}.");
+        }
+    }
+}
```
Comment (metalgearsloth) [inline Content.Client/_CD/WelcomeMessage/WelcomeWindow.xaml.cs:42]: ReadToEndAsync?

File: Content.Client/_CD/WelcomeMessage/WelcomeWindowUIController.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        _window?.Open();
+    }
+
+    private void OnDoNotShow()
+    {
+        _window?.Close();
+        _resource.UserData.WriteAllText(new ResPath(LastSeenRevisionPath), TextRevision.ToString());
+    }
+}
```
[the remark this answers is not in the document]
Comment (dffdff2423) [reply Content.Client/_CD/WelcomeMessage/WelcomeWindowUIController.cs:64]: I will look into it, I mostly just copied what the changelog was doing.

File: Content.Client/_CD/WelcomeMessage/WelcomeWindow.xaml.cs (hunk 3 above)
Comment (dffdff2423) [reply Content.Client/_CD/WelcomeMessage/WelcomeWindow.xaml.cs]: I was unable to figure out how to get that to work (I have only used c# for ss14). 

My attempt consisted of changing LoadContents to async (returning a Task) and awaiting `file.ReadToEndAsync`. I then added an optional Task field to the UI controller and called LoadContents with this snippet:

```csharp
            _loadTask = _window.LoadContents(new ResPath(ContentsPath))
                .ContinueWith(async res =>
            {
                _currentHash = await res;
            });
```

Finally, I waiting on _loadTask before comparing the hashes.

This implementation seemed to cause the game to hang most of the time upon loading in.

Comment (dffdff2423) [reply Content.Client/_CD/WelcomeMessage/WelcomeWindow.xaml.cs]: Is there an example in ss14 that you can point me to? It looks like most other things that read files use the non-async ReadToEnd.

**Comment by dffdff2423** [comment]
I am going to close this PR because this is not a particularly elegant solution and I am not entirely sure why the test is failing. I am not likely to revisit this for a while, if someone else wants to implement this a better way would be to automatically pop open the guidebook with a welcome article on the first join.

**Pull request closed without merging**